Fix crash on Android devices supporting S3TC

lawnjelly opened the pull request:
Fixes #28308

Android devices that support S3TC will currently crash if you export / run from the IDE a project that has S3TC textures imported. This is because these are not exported in the APK (android exporter, exporter.cpp, get_preset_features()), but the .import file still contains a reference to the S3TC files, and the app will attempt to load the resource and fail.

This PR fixes this by simply faking lack of hardware support for S3TC on android devices.

An alternative method maybe to remove the reference to the s3tc in the .import file.

lawnjelly commented:
Note there may be a more elegant way of solving this, but I've been using this fix for the past few months. See the issue quoted above for details.

akien-mga commented:
That's hacky, the proper fix would be to prevent trying to load S3TC files, or remove references to them when exporting for a platform which doesn't use them.

lawnjelly commented:
> That's hacky, the proper fix would be to prevent trying to load S3TC files, or remove references to them when exporting for a platform which doesn't use them.

Yep very true and I mentioned this in the original issue, I'm really just being cheeky and using the (hacky) PR to try and stimulate some direction on the best way to address this from some of the more core guys, maybe in one of the PR meetings (otherwise I'm afraid it will languish in issue purgatory, and it does and will cause crashes on affected hardware with this and similar bugs, because it might be a bit of a general oversight in the export / loading logic).

I think the removing references at export would be ideal but a little bug prone (being a text file), and finding the loading bit was a bit impenetrable when I looked at the time, but I'll have another look.

If this is bad form I can close this and try and get some advice on the irc channel. :+1:

akien-mga commented:
> I'm really just being cheeky and using the (hacky) PR to try and stimulate some direction on the best way to address this from some of the more core guys, maybe in one of the PR meetings (otherwise I'm afraid it will languish in issue purgatory

Hehe that's a proven method indeed, a better PR is often better to build momentum around an issue than just commenting on one issue out of 5000. :)

m4gr3d reviewed (commented):
Agree it's not ideal, but should it be considered as a temp fix for 3.2?

I lack expertise in the export and loading logic so it's possible that fixing these instead may be simple and quick, but if not, this fix may do.

lawnjelly commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -5612,6 +5612,15 @@ void RasterizerStorageGLES2::initialize() {
+	// S3TC it will crash trying to load these textures, as they are not exported in the APK. This is a simple way
+	// to prevent Android devices trying to load S3TC, by faking lack of hardware support.
+#ifndef TOOLS_ENABLED
+#if defined ANDROID_ENABLED
```
Sorry I've been away from internet (and will be until at least tomorrow afternoon I expect). Yup, I did notice the potential for the same issue to occur on iOS and javascript, although I can't confirm or not whether those do occur in practice (as iOS is quite fixed hardware, and I haven't been able to get an HTML5 export working for quite some time!).

The quick fix here does work for this specific case, and according to some hardware stats I looked at, it might affect up to 1/5 of android devices (those that support S3TC, and they would probably also crash with this). It is possible to workaround by unticking S3TC in the compressed texures bit of project settings, and reimporting textures before making the final export for android. However this is easy to forget, and not at all obvious. So it is potentially quite an easy way of expanding the compatibility with Android devices for 3.2.

However I am in agreement that it is a bit hacky and is potentially the symptom of a larger problem whereby the features in the export are not taken into account at runtime, it's something that we (or probably me lol) need to ask reduz (is it likely to be a one off bug, or a systemic problem) either in PR meeting or on IRC.

akien-mga replied:
For Javascript it's more complex than that, as it does use S3TC on desktop browsers, but likely something else on mobile browsers (didn't check what exactly).

lawnjelly replied:
Rather than modifying the import files, it might be easier to use the approach of saving which features have successfully been exported (maybe this is done already? A long time since I looked at this), loading this feature set in at runtime and perhaps converting to bitflags, then checking the bitflags before attempting to load a resource. This would solve the problem in a generic way as it could potentially occur in several areas.

Or we could simply log a warning message when failing to load a resource, and go for the next one in the list, rather than crashing... :smile: 

I'll try and have another look at this problem after the weekend.

aaronfranke commented:
@lawnjelly Is this still desired? If so, it needs to be rebased on the latest master branch and made to work for Vulkan.

lawnjelly commented:
> @lawnjelly Is this still desired? If so, it needs to be rebased on the latest master branch and made to work for Vulkan.

Yes, we still need a fix for this, as it will be causing crashes in devices. We really need a discussion of this problem in PR meeting (as it may be a common issue in the export process).

TIBI4 commented:
**A workaround as I'm working on 3.2.3 stable official.**  _(this worked for me)_

In **Project** -> **Export...** -> **Your Android Preset To Use** -> **Features** -> **Custom**:

Add **"s3tx"** to the custom list.
![image](https://user-images.githubusercontent.com/25259582/96217037-2ce9f180-0f58-11eb-8e12-9e555899b9e0.png)

The project should start normally. **But I have a question**, as I don't know what is happening there though. **Is this a bad idea?**

lawnjelly commented:
That's an interesting approach. The downside is that if it is including the s3tc files, the apk is bigger than it could be (as presumably you are also including ETC for the majority of devices that don't support s3tc).

I haven't tested this, but there may be another possible workaround:

* If you go to the `project_settings->rendering->vram_compression` page, and click to select import_s3tc, in the top right of the dialog you can add an override for android, and turn off import_s3tc for android only.

I've no idea whether this would work, not having time to test at the moment.

akien-mga commented:
After thinking some more about it, I think the approach taken in this PR might actually be good.

In theory those Android devices could use and benefit from having S3TC, which would imply register `s3tc` as a supported feature for Android as discussed above. But the main issue is that many devices *do not* support S3TC, and it becomes tricky to define what APK to provide to users of each type of device.

So as long as all Android devices support ETC2, it's likely best to intentionally ignore S3TC support on Android and mandate the use of ETC2.

akien-mga commented:
> * If you go to the `project_settings->rendering->vram_compression` page, and click to select import_s3tc, in the top right of the dialog you can add an override for android, and turn off import_s3tc for android only.

I don't think that would work, it would disable *importing* S3TC when running the editor on an Android device, but the import would still happen when you develop on PC.

lawnjelly commented:
I've (rather traumatically :grin: ) rebased the PR, although it is on 3.2. I can't really make PRs on master at the moment as I get vulkan crashes, however this should be quite easy to copy paste there.

Although there should really be a better longterm solution, it's been a year since I made the PR so I'm of the view we should probably fix this as a stopgap at least, as it will be causing crashes. We can easily replace it if we get an overhaul of this whole import area.

lawnjelly commented:
Just for future reference here is discussion on it today:

```
<Akien> reduz: can we discuss #32255? poke lawnjelly
<IssueBot> #32255: Fix crash on Android devices supporting S3TC | https://git.io/JesWs
<Akien> It seemed hacky to me initially but now I'm thinking this might well be the cleanest way to do it if we don't want to complexify the distribution of Android games
<reduz> Akien: could be fixed by changing the save order, by saving etc first and s3tc second
<Akien> Outright ignoring S3TC support might be better than having some devices than needed S3TC and others that don't, so you either need 2 APKs (and a way to differentiate) or impose the size of S3TC imports on all devices
<reduz> though we may have the opposite problem on PC, as some do support ETC
<reduz> Akien: maybe we should just fail if the file is not found and go to the next one
<reduz> Akien: or indeed, remove it from the ,import file on export is likely best
<lawnjelly> back. am here.
<Akien> reduz: I guess we'd need to do both, with an export preset setting to control what format you want to export
<lawnjelly> I was wondering if it was a general problem with export.
<Akien> So if you know you actually want s3tc on android, you could still have a way to keep them.
<lawnjelly> The exporters export only a subset of what is available
<lawnjelly> but it is all mentioned in the import files.
<lawnjelly> Whether there was a general solution, not just for s3tc.
<Akien> Yeah the export process should likely remove the unsupported formats from .import when exporting
<lawnjelly> Could that be less future proof than including the info as to which are exported?
<lawnjelly> Just to play devil's advocate
<Akien> reduz: is there already code that decides what imported textures to include in the PCK based on feature flags?
<lawnjelly> every time we added a new feature, we might have to change the exporters in that case. Whereas with the feature flags / list of features there would be no future changes needed.
<Akien> Currently the default feature flags for Android are here: https://github.com/godotengine/godot/blob/3.2/platform/android/export/export.cpp#L1631-L1647
<Akien> s3tc is not included, so I guess it's not exported due to this feature flag. Which also means that if you add the s3tc feature flag manually in your preset, then it should be exported.
<reduz> Akien: yes, its in the exporter
<Akien> So removing references to what is not being exported (also based on feature flags) should work fine I guess.
<reduz> yes its kind of easy you use ConfigFile API to modify it
<Akien> lawnjelly: we phrase it differently but I think we all agree :D
<lawnjelly> Ah that's far easier than editing the text manually.
<lawnjelly> Yeah that could be a good solution.
<lawnjelly> :)
<lawnjelly> There might be someone more expert with the exporters who is better placed to do this than me
<Akien> Here's the relevant code I think: https://github.com/godotengine/godot/blob/3.2/editor/editor_export.cpp#L775-L835
<lawnjelly> But if not I can try and do it
```

DmitriySalnikov commented:
Unfortunately, this problem applies to the new `Windows Subsystem for Android`. Therefore, to test apps with VRAM textures, it is necessary to add the `s3tc` feature.
I tested it in `tps-demo` and in `platformer-3d` in `GLES3` mode (although here the fix is only for `GLES2`).
The `tps-demo` takes a very long time to load, shows 3-30 FPS, and the size of the game has increased almost 2 times, but it works.

What about giving developers the option to add etc2, s3tc, pvrtc to the archive at the same time, but in AAB separate these textures for target devices?

<details>
<summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/7782218/144551757-0f491ddd-979a-4abe-bf0b-7feb26c12d1a.png)

![image](https://user-images.githubusercontent.com/7782218/144549633-5a94cfce-d037-47c4-9e55-3ab18435e6df.png)

![image](https://user-images.githubusercontent.com/7782218/144549643-bb062cc5-798a-4b1c-8444-94ddd4394a63.png)

![image](https://user-images.githubusercontent.com/7782218/144551699-944016ae-fb59-4ea9-b243-19d101186c46.png)


</details>

Calinou commented:
> What about giving developers the option to add etc2, s3tc, pvrtc to the archive at the same time, but in AAB separate these textures for target devices?

As I understand it, PVRTC compression isn't really used much anymore due to its low quality and it only working on PowerVR GPUs (i.e. iOS devices).

DmitriySalnikov commented:
@Calinou but there are still many devices with Adreno graphics and `ATC` compression. There is also an `ASTC`.
But at the moment there is only `s3tc`, `etc`, `etc2` in Godot.
To be honest, I don't mind if `etc2` is forced.

Gromph commented:
We had a user running the Android version of our game on an intel chromebook report this crash.  This Pull Request fixed it. Thank you.

akien-mga commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -5943,6 +5943,13 @@ void RasterizerStorageGLES2::initialize() {
+	// If the desktop build is using S3TC, and you export / run from the IDE for android, if the device supports
+	// S3TC it will crash trying to load these textures, as they are not exported in the APK. This is a simple way
+	// to prevent Android devices trying to load S3TC, by faking lack of hardware support.
+#if !defined(TOOLS_ENABLED) && defined(ANDROID_ENABLED)
```
As discussed, you could likely remove `TOOLS_ENABLED` and add `IPHONE_ENABLED` for good measure, just in case (probably no device supporting it but who knows).

lawnjelly committed:
Fix for S3TC on Android and IOS devices

On some platforms, exporters are prevented from exporting S3TC textures. This causes problems if the .import file contains a reference to such a texture - the exported project will attempt to load the S3TC, fail, and probably crash.

This PR prevents this problem by faking lack of hardware support for S3TC on the affected platforms. This prevents the engine attempting to load the S3TC and avoids the problem.

akien-mga commented:
Thanks! Took a while to assess and come to the conclusion that we don't have the time/will to do this better for now :D

akien-mga commented:
Cherry-picked for 3.4.3.
